File: src/bin/day2_2.rs
use std::fs::File;
use std::io::{BufReader, BufRead};

const ROCK: usize = 1;
const PAPER: usize = 2;
const SIZ: usize = 3;

fn get_score_of_shape(shape: &str)-> usize{
    if shape == "A" || shape == "X" {
	ROCK
    }else if shape == "B" || shape == "Y" {
	PAPER
    }else if shape == "C" || shape == "Z" {
	SIZ
    }else {
	panic!("Should never reach here");
    }
}

fn get_loosing_shape(shape: &str)->String{
    if shape == "A" {
	return "C".to_string();
    }
    if shape == "B" {
	return "A".to_string();
    }
    if shape == "C" {
	return "B".to_string();
    }
    return "-".to_string();//feed back failure
}

fn get_drawing_shape(shape: &str)-> String{
    return shape.clone().to_string()
}

fn get_winning_shape(shape: &str)-> String{
    if shape == "A" {
	return "B".to_string();
    }
    if shape == "B" {
	return "C".to_string();
    }
    if shape == "C" {
	return "A".to_string();
    }
    return "-".to_string();//feed back failure
}

fn get_score(them: &str, me: &str)-> usize{
    if me == "X" {
	let loose = get_loosing_shape(them);
	return 0+get_score_of_shape(&loose);
    }
    if me == "Y" {
	let draw = get_drawing_shape(them);
	return 3+get_score_of_shape(&draw);
    }
    if me == "Z" {
	let win = get_winning_shape(them);
	return 6+get_score_of_shape(&win);
    }
    println!("WTF!!! {} {}",me, them);
    0
}

fn main(){
    let mut score: usize = 0;
    for l in BufReader::new(File::open("./day2_1.txt").unwrap()).lines(){
	let line = l.unwrap();
	let vals = line.split(' ').collect::<Vec<&str>>();
	let (them,me) = (vals[0],vals[1]);
	score+=get_score(them,me);
    }
    println!("Total: {score}");
}

File: src/day2_2.zig
const std = @import("std");
const utils = @import("utils");
const ArrayList = std.ArrayList;

// Rock > Siz
// Siz > Paper
// Paper > Rock

// A = Rock = 1 = X
// B = Paper = 2 = Y
// C = Siz = 3 = Z

// Loss = 0
// Win = 6
// Draw = 3

fn scoreOfChoice(choice: u8) usize{
    switch(choice){
        88 => return 1,
        89 => return 2,
        90 => return 3,
        else => unreachable
    }
}

fn getScoreForLoss(op: u8) usize {
    switch(op){
        65 => return scoreOfChoice(90),
        66 => return scoreOfChoice(88),
        67 => return scoreOfChoice(89),
        else => unreachable
    }
}

fn getScoreForDraw(op: u8) usize {
    switch(op){
        65 => return scoreOfChoice(88),
        66 => return scoreOfChoice(89),
        67 => return scoreOfChoice(90),
        else => unreachable
    }
}


fn getScoreForWin(op: u8) usize {
    switch(op){
        65 => return scoreOfChoice(89),
        66 => return scoreOfChoice(90),
        67 => return scoreOfChoice(88),
        else => unreachable
    }
}

fn getScore(op: u8, me: u8) usize {
    switch(me){
        88 => return getScoreForLoss(op),
        89 => return 3 + getScoreForDraw(op),
        90 => return 6 + getScoreForWin(op),
        else => unreachable
    }
}

pub fn main() !void{
    const stdout = std.io.getStdOut().writer();
    try stdout.print("Day 2 p1 solution\n", .{});
    var lines: ArrayList(ArrayList(u8)) = try utils.readLinesFromFile(100, "res/day2_1.txt");
    defer lines.deinit();

    var score: usize = 0;

    for(lines.items)|line|{
        if(line.items.len==0){
            continue;
        }
        const op = line.items[0];
        const me = line.items[2];
        const result_score = getScore(op,me);
        score += result_score;
    }
    try stdout.print("Total score is {d}\n",.{score});
}
